File: 2025_Github_Screener.rb
# frozen_string_literal: true

require 'octokit'
require 'base64'
require 'set'
require 'yaml' # Hinzugefügt für die Konfigurationsdatei

# ===================================================================
# CONFIGURATION MANAGEMENT
# ===================================================================

CONFIG_FILE = 'config.yml'

# Lädt die Konfiguration aus der YAML-Datei.
# Wenn die Datei nicht existiert oder unvollständig ist, wird eine interaktive Einrichtung gestartet.
def load_or_create_config
  config = if File.exist?(CONFIG_FILE)
             YAML.load_file(CONFIG_FILE) || {}
           else
             {}
           end

  if config['GITHUB_USER'].nil? || config['GITHUB_USER'] == 'USER_NAME' || config['ACCESS_TOKEN'].nil? || config['ACCESS_TOKEN'] == 'YOUR_TOKEN'
    puts "Konfiguration nicht gefunden oder unvollständig. Bitte jetzt einrichten."
    
    print "Gib deinen GitHub-Benutzernamen ein: "
    config['GITHUB_USER'] = gets.chomp
    
    print "Gib dein GitHub Personal Access Token ein: "
    config['ACCESS_TOKEN'] = gets.chomp

    File.write(CONFIG_FILE, config.to_yaml)
    puts "✅ Konfiguration wurde in '#{CONFIG_FILE}' gespeichert.\n"
  end
  
  config
end

# Lade die Konfiguration
config = load_or_create_config

# Weise die Konfigurationswerte den Konstanten zu.
# Das Umgebungstoken hat weiterhin Vorrang.
ACCESS_TOKEN = ENV['GITHUB_TOKEN'] || config['ACCESS_TOKEN']
GITHUB_USER = config['GITHUB_USER']

# Stop-Wörter bleiben unverändert
STOP_WORDS = Set.new(%w(
  https http href com github bash www org de a about an and are as at be by for from how i in is it of on or that the this to was what when where who will with the to and of a in is it for on that as with he she they we me you my your our I be have do not have was were if then else for while do code file files gem build setup config run installation usage license mit gpl data lib docs new get use using via from this that these those an example examples please feel free to more also just like some any all its can not will readme md out there because been through into only repo repository project projects app application service api client server test tests feature features version update release change fix add remove refactor style docs chore ci build performance security support help contact information details note important))

# ===================================================================
# SCRIPT LOGIC (unverändert)
# ===================================================================

puts "Initializing screener for user: #{GITHUB_USER}..."

# Authenticate the client with the token
client = Octokit::Client.new(access_token: ACCESS_TOKEN)

# Make sure the user exists
begin
  user = client.user(GITHUB_USER)
rescue Octokit::NotFound
  puts "Error: User '#{GITHUB_USER}' not found."
  exit
rescue Octokit::Unauthorized
  puts "Error: GitHub Token ist ungültig oder hat nicht die nötigen Rechte."
  exit
end

# Initialize counters and collectors
total_lines_of_code = 0
total_lines_of_docs = 0
total_commits = 0
language_stats = Hash.new(0)
repo_names = []
first_activity_date = nil
last_activity_date = nil
keyword_counts = Hash.new(0)

# Helper function for keyword analysis
def process_and_count_words(text, counts)
  return unless text && !text.empty?
  words = text.downcase.scan(/[a-z0-9]+/) # Extract words/numbers
  words.each do |word|
    next if STOP_WORDS.include?(word) || word.length < 3
    counts[word] += 1
  end
end

# Fetch all public repositories for the user
puts "Fetching all public repositories..."
repos = client.repos(GITHUB_USER)

puts "Starting analysis of #{repos.size} repositories. This may take a while..."
puts "---"

# Iterate through each repository
repos.each_with_index do |repo, index|
  full_repo_name = repo.full_name
  puts "[#{index + 1}/#{repos.size}] Analyzing: #{full_repo_name}"
  
  # 1. Collect repo name
  repo_names << full_repo_name

  # 2. Track first and last activity dates
  if first_activity_date.nil? || repo.created_at < first_activity_date
    first_activity_date = repo.created_at
  end
  if last_activity_date.nil? || repo.pushed_at > last_activity_date
    last_activity_date = repo.pushed_at
  end
  
  # 3. Process keywords from repo name
  process_and_count_words(repo.name.gsub(/[-_]/, ' '), keyword_counts)
  
  # 4. Analyze repo contents
  begin
    # Language statistics
    repo_langs = client.languages(full_repo_name)
    repo_langs.to_h.each { |lang, bytes| language_stats[lang] += bytes }

    # README analysis (docs lines and keywords)
    begin
      readme = client.readme(full_repo_name)
      readme_content = Base64.decode64(readme.content)
      total_lines_of_docs += readme_content.lines.count
      process_and_count_words(readme_content, keyword_counts)
    rescue Octokit::NotFound
      # No README found
    end

    # Lines of code count
    last_commit_sha = client.branch(full_repo_name, repo.default_branch).commit.sha
    tree = client.tree(full_repo_name, last_commit_sha, recursive: true).tree
    tree.each do |file|
      next unless file.type == 'blob' && file.size > 0 && file.size < 1_000_000 # Limit: 1MB
      begin
        blob = client.blob(full_repo_name, file.sha)
        if blob.encoding == 'base64' && (content = Base64.decode64(blob.content)).valid_encoding?
          total_lines_of_code += content.lines.count
        end
      rescue Octokit::Error
        # Blob could not be fetched
      end
    end

    # Commit activity
    repo_commits = client.commit_activity_stats(full_repo_name)
    total_commits += repo_commits.map(&:total).sum if repo_commits
  rescue Octokit::Error => e
    puts "Could not fully analyze repository '#{full_repo_name}': #{e.message}"
  end
end

puts "\n---"
puts "✅ Analysis complete!"
puts "---"

# Format and print the results
puts "\n📊 Results for #{user.name} (@#{GITHUB_USER})"
puts "=============================================="

# Activity Dates
puts "\n🗓️ Activity Span"
puts "First activity (oldest repo): #{first_activity_date&.strftime('%Y-%m-%d') || 'N/A'}"
puts "Last activity (latest push):  #{last_activity_date&.strftime('%Y-%m-%d') || 'N/A'}"

# General Stats
puts "\n📈 General Statistics"
puts "Total lines of code: #{total_lines_of_code.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse}"
puts "Total documentation lines (in READMEs): #{total_lines_of_docs.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse}"
puts "Total commits (last year): #{total_commits.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse}"

# Top 10 Keywords
puts "\n🔑 Top 10 Keywords (from Repo Names & READMEs)"
if keyword_counts.empty?
  puts "No keywords found."
else
  top_10_keywords = keyword_counts.sort_by { |_, count| -count }.first(10)
  top_10_keywords.each_with_index do |(word, count), i|
    puts "#{i + 1}. #{word} (#{count} mentions)"
  end
end

# Programming Languages
puts "\n💻 Programming Languages Used:"
total_bytes = language_stats.values.sum
if total_bytes > 0
  sorted_langs = language_stats.sort_by { |_, bytes| -bytes }
  sorted_langs.each do |lang, bytes|
    percentage = (bytes.to_f / total_bytes * 100).round(2)
    puts "   - #{lang}: #{percentage}%" if percentage > 0.1
  end
else
  puts "No language data found."
end

# Repository List
puts "\n📚 Repository List (#{repo_names.size} total)"
repo_names.each { |name| puts "   - #{name}" }

puts "=============================================="

File: 2025_Github_Screener.py
import os
import sys
import base64
import re
from collections import Counter

import yaml
from github import Github, GithubException

# ===================================================================
# CONFIGURATION
# ===================================================================

CONFIG_FILE = 'config.yml'

def load_or_create_config():
    """Lädt die Konfiguration oder startet die interaktive Einrichtung."""
    config = {}
    if os.path.exists(CONFIG_FILE):
        with open(CONFIG_FILE, 'r') as f:
            config = yaml.safe_load(f) or {}

    if not config.get('GITHUB_USER') or config.get('GITHUB_USER') == 'USER_NAME' or not config.get('ACCESS_TOKEN') or config.get('ACCESS_TOKEN') == 'YOUR_TOKEN':
        print("Konfiguration nicht gefunden oder unvollständig. Bitte jetzt einrichten.")
        
        config['GITHUB_USER'] = input("Gib deinen GitHub-Benutzernamen ein: ")
        config['ACCESS_TOKEN'] = input("Gib dein GitHub Personal Access Token ein: ")

        with open(CONFIG_FILE, 'w') as f:
            yaml.dump(config, f)
        print(f"✅ Konfiguration wurde in '{CONFIG_FILE}' gespeichert.\n")
    
    return config

# Stop-Wörter, um relevantere Ergebnisse zu erhalten.
STOP_WORDS = {
    'https', 'http', 'href', 'com', 'github', 'bash', 'www', 'org', 'de', 'a', 'about', 'an', 'and', 'are', 
    'as', 'at', 'be', 'by', 'for', 'from', 'how', 'i', 'in', 'is', 'it', 'of', 'on', 'or', 'that', 'the', 
    'this', 'to', 'was', 'what', 'when', 'where', 'who', 'will', 'with', 'he', 'she', 'they', 'we', 'me', 
    'you', 'my', 'your', 'our', 'do', 'not', 'have', 'were', 'if', 'then', 'else', 'while', 'code', 'file', 
    'files', 'gem', 'build', 'setup', 'config', 'run', 'installation', 'usage', 'license', 'mit', 'gpl', 
    'data', 'lib', 'docs', 'new', 'get', 'use', 'using', 'via', 'from', 'these', 'those', 'example', 
    'examples', 'please', 'feel', 'free', 'more', 'also', 'just', 'like', 'some', 'any', 'all', 'its', 
    'can', 'readme', 'md', 'out', 'there', 'because', 'been', 'through', 'into', 'only', 'repo', 
    'repository', 'project', 'projects', 'app', 'application', 'service', 'api', 'client', 'server', 
    'test', 'tests', 'feature', 'features', 'version', 'update', 'release', 'change', 'fix', 'add', 
    'remove', 'refactor', 'style', 'chore', 'ci', 'performance', 'security', 'support', 'help', 
    'contact', 'information', 'details', 'note', 'important'
}

# ===================================================================

def process_and_count_words(text, counter):
    """Extrahiert und zählt Wörter aus einem Text."""
    if not text:
        return
    words = re.findall(r'[a-z0-9]+', text.lower())
    for word in words:
        if word not in STOP_WORDS and len(word) >= 3:
            counter[word] += 1

def main():
    """Hauptfunktion des Skripts."""
    config = load_or_create_config()
    
    ACCESS_TOKEN = os.environ.get('GITHUB_TOKEN') or config['ACCESS_TOKEN']
    GITHUB_USER = config['GITHUB_USER']

    print(f"Initializing screener for user: {GITHUB_USER}...")

    # Authentifiziere den Client
    g = Github(ACCESS_TOKEN)

    # Stelle sicher, dass der Benutzer existiert
    try:
        user = g.get_user(GITHUB_USER)
    except GithubException as e:
        if e.status == 404:
            print(f"Error: User '{GITHUB_USER}' not found.")
        elif e.status == 401:
            print("Error: GitHub Token ist ungültig oder hat nicht die nötigen Rechte.")
        else:
            print(f"An unexpected error occurred: {e}")
        sys.exit()

    # Initialisiere Zähler und Sammlungen
    total_lines_of_code = 0
    total_lines_of_docs = 0
    total_commits = 0
    language_stats = Counter()
    repo_names = []
    first_activity_date = None
    last_activity_date = None
    keyword_counts = Counter()

    # Hole alle öffentlichen Repositories
    print("Fetching all public repositories...")
    repos = list(user.get_repos())
    repo_count = len(repos)
    print(f"Starting analysis of {repo_count} repositories. This may take a while...")
    print("---")

    # Iteriere durch jedes Repository
    for i, repo in enumerate(repos):
        print(f"[{i + 1}/{repo_count}] Analyzing: {repo.full_name}")
        
        # 1. Sammle Repo-Namen
        repo_names.append(repo.full_name)

        # 2. Verfolge erstes und letztes Aktivitätsdatum
        if first_activity_date is None or repo.created_at < first_activity_date:
            first_activity_date = repo.created_at
        if last_activity_date is None or repo.pushed_at < last_activity_date:
            last_activity_date = repo.pushed_at
        
        # 3. Verarbeite Schlüsselwörter aus dem Repo-Namen
        process_and_count_words(re.sub(r'[-_]', ' ', repo.name), keyword_counts)

        try:
            # 4. Analysiere den Inhalt des Repos
            # Sprachstatistiken
            for lang, byte_count in repo.get_languages().items():
                language_stats[lang] += byte_count

            # README-Analyse
            try:
                readme = repo.get_readme()
                readme_content = base64.b64decode(readme.content).decode('utf-8', errors='ignore')
                total_lines_of_docs += len(readme_content.splitlines())
                process_and_count_words(readme_content, keyword_counts)
            except GithubException:
                pass # Kein README gefunden

            # Zeilen Code zählen
            tree = repo.get_git_tree(repo.default_branch, recursive=True).tree
            for element in tree:
                if element.type == 'blob' and 0 < element.size < 1_000_000:
                    try:
                        blob_content = repo.get_git_blob(element.sha).content
                        decoded_content = base64.b64decode(blob_content).decode('utf-8', errors='ignore')
                        total_lines_of_code += len(decoded_content.splitlines())
                    except Exception:
                        pass # Konnte Blob nicht verarbeiten

            # Commit-Aktivität (letztes Jahr)
            commit_stats = repo.get_stats_commit_activity()
            if commit_stats:
                total_commits += sum(stat.total for stat in commit_stats)

        except GithubException as e:
            print(f"Could not fully analyze repository '{repo.full_name}': {e.data.get('message', 'API Error')}")

    print("\n---")
    print("✅ Analysis complete!")
    print("---")

    # Ergebnisse formatieren und ausgeben
    print(f"\n📊 Results for {user.name} (@{GITHUB_USER})")
    print("==============================================")

    print("\n🗓️ Activity Span")
    print(f"First activity (oldest repo): {first_activity_date.strftime('%Y-%m-%d') if first_activity_date else 'N/A'}")
    print(f"Last activity (latest push):  {last_activity_date.strftime('%Y-%m-%d') if last_activity_date else 'N/A'}")

    print("\n📈 General Statistics")
    print(f"Total lines of code: {total_lines_of_code:,}")
    print(f"Total documentation lines (in READMEs): {total_lines_of_docs:,}")
    print(f"Total commits (last year): {total_commits:,}")

    print("\n🔑 Top 10 Keywords (from Repo Names & READMEs)")
    if not keyword_counts:
        print("No keywords found.")
    else:
        for i, (word, count) in enumerate(keyword_counts.most_common(10)):
            print(f"{i + 1}. {word} ({count} mentions)")

    print("\n💻 Programming Languages Used:")
    total_bytes = sum(language_stats.values())
    if total_bytes > 0:
        sorted_langs = sorted(language_stats.items(), key=lambda item: item[1], reverse=True)
        for lang, byte_count in sorted_langs:
            percentage = (byte_count / total_bytes * 100)
            if percentage > 0.1:
                print(f"   - {lang}: {percentage:.2f}%")
    else:
        print("No language data found.")

    print(f"\n📚 Repository List ({len(repo_names)} total)")
    for name in repo_names:
        print(f"   - {name}")

    print("==============================================")


if __name__ == '__main__':
    main()